Support Groovy 4

Currently, this library builds with Groovy 2.4 and works fine with Groovy 3.x. However, with Groovy 4, it fails with a `ChangeLogParseException` with text `Unrecognized root element`.

For instance, given the following script:
```groovy
databaseChangeLog {
    changeSet(id: '123', author: 'me') {
        createTable(tableName: 'my_table') {
            column(name: 'id', type: 'BIGINT', autoIncrement: true) {
                constraints(primaryKey: true, primaryKeyName: 'id_PK')
            }
        }
    }
}
```

Running with Groovy 4, it fails with: `Caused by: liquibase.exception.ChangeLogParseException: Unrecognized root element changeSet`. It works with Groovy 3.

<details>
<summary style="font-size:14px">Full stack trace (from a Spring Boot application)</summary>
<p>

```
Caused by: liquibase.exception.ChangeLogParseException: Unrecognized root element changeSet
	at jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method) ~[?:?]
	at jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62) ~[?:?]
	at jdk.internal.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45) ~[?:?]
	at java.lang.reflect.Constructor.newInstance(Constructor.java:490) ~[?:?]
	at org.codehaus.groovy.reflection.CachedConstructor.invoke(CachedConstructor.java:73) ~[groovy-4.0.5.jar:4.0.5]
	at org.codehaus.groovy.reflection.CachedConstructor.doConstructorInvoke(CachedConstructor.java:60) ~[groovy-4.0.5.jar:4.0.5]
	at org.codehaus.groovy.runtime.callsite.ConstructorSite$ConstructorSiteNoUnwrap.callConstructor(ConstructorSite.java:86) ~[groovy-4.0.5.jar:4.0.5]
	at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCallConstructor(CallSiteArray.java:57) ~[groovy-4.0.5.jar:4.0.5]
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.callConstructor(AbstractCallSite.java:263) ~[groovy-4.0.5.jar:4.0.5]
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.callConstructor(AbstractCallSite.java:277) ~[groovy-4.0.5.jar:4.0.5]
	at liquibase.parser.ext.GroovyLiquibaseChangeLogParser$_getChangeLogMethodMissing_closure3.doCall(GroovyLiquibaseChangeLogParser.groovy:94) ~[liquibase-groovy-dsl-3.0.2.jar:?]
	at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:?]
	at jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[?:?]
	at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:?]
	at java.lang.reflect.Method.invoke(Method.java:566) ~[?:?]
	at org.codehaus.groovy.reflection.CachedMethod.invoke(CachedMethod.java:343) ~[groovy-4.0.5.jar:4.0.5]
	at org.codehaus.groovy.runtime.metaclass.ClosureMetaMethod.invoke(ClosureMetaMethod.java:88) ~[groovy-4.0.5.jar:4.0.5]
	at groovy.lang.MetaClassImpl.invokeMissingMethod(MetaClassImpl.java:926) ~[groovy-4.0.5.jar:4.0.5]
	at groovy.lang.MetaClassImpl.invokePropertyOrMissing(MetaClassImpl.java:1428) ~[groovy-4.0.5.jar:4.0.5]
	at groovy.lang.MetaClassImpl.doInvokeMethod(MetaClassImpl.java:1350) ~[groovy-4.0.5.jar:4.0.5]
	at groovy.lang.MetaClassImpl.invokeMethod(MetaClassImpl.java:1103) ~[groovy-4.0.5.jar:4.0.5]
	at groovy.lang.ExpandoMetaClass.invokeMethod(ExpandoMetaClass.java:1142) ~[groovy-4.0.5.jar:4.0.5]
	at groovy.lang.MetaClassImpl.invokeMethod(MetaClassImpl.java:1009) ~[groovy-4.0.5.jar:4.0.5]
	at org.codehaus.groovy.runtime.InvokerHelper.invokePogoMethod(InvokerHelper.java:620) ~[groovy-4.0.5.jar:4.0.5]
	at org.codehaus.groovy.runtime.InvokerHelper.invokeMethod(InvokerHelper.java:603) ~[groovy-4.0.5.jar:4.0.5]
	at org.codehaus.groovy.runtime.metaclass.ClosureMetaClass.invokeOnDelegationObjects(ClosureMetaClass.java:391) ~[groovy-4.0.5.jar:4.0.5]
	at org.codehaus.groovy.runtime.metaclass.ClosureMetaClass.invokeMethod(ClosureMetaClass.java:328) ~[groovy-4.0.5.jar:4.0.5]
	at groovy.lang.MetaClassImpl.invokeMethod(MetaClassImpl.java:1009) ~[groovy-4.0.5.jar:4.0.5]
	at org.codehaus.groovy.vmplugin.v8.IndyInterface.fromCache(IndyInterface.java:318) ~[groovy-4.0.5.jar:4.0.5]
	at Script1$_run_closure1.doCall(Script1.groovy:2) ~[?:?]
	at Script1$_run_closure1.doCall(Script1.groovy) ~[?:?]
	at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:?]
	at jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[?:?]
	at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:?]
	at java.lang.reflect.Method.invoke(Method.java:566) ~[?:?]
	at org.codehaus.groovy.reflection.CachedMethod.invoke(CachedMethod.java:343) ~[groovy-4.0.5.jar:4.0.5]
	at groovy.lang.MetaMethod.doMethodInvoke(MetaMethod.java:328) ~[groovy-4.0.5.jar:4.0.5]
	at org.codehaus.groovy.runtime.metaclass.ClosureMetaClass.invokeMethod(ClosureMetaClass.java:279) ~[groovy-4.0.5.jar:4.0.5]
	at groovy.lang.MetaClassImpl.invokeMethod(MetaClassImpl.java:1009) ~[groovy-4.0.5.jar:4.0.5]
	at org.codehaus.groovy.runtime.callsite.PogoMetaClassSite.call(PogoMetaClassSite.java:39) ~[groovy-4.0.5.jar:4.0.5]
	at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCall(CallSiteArray.java:45) ~[groovy-4.0.5.jar:4.0.5]
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:125) ~[groovy-4.0.5.jar:4.0.5]
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:130) ~[groovy-4.0.5.jar:4.0.5]
	at liquibase.parser.ext.GroovyLiquibaseChangeLogParser.processDatabaseChangeLogRootElement(GroovyLiquibaseChangeLogParser.groovy:134) ~[liquibase-groovy-dsl-3.0.2.jar:?]
	at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:?]
	at jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[?:?]
	at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:?]
	at java.lang.reflect.Method.invoke(Method.java:566) ~[?:?]
	at org.codehaus.groovy.reflection.CachedMethod.invoke(CachedMethod.java:343) ~[groovy-4.0.5.jar:4.0.5]
	at groovy.lang.MetaMethod.doMethodInvoke(MetaMethod.java:328) ~[groovy-4.0.5.jar:4.0.5]
	at org.codehaus.groovy.runtime.metaclass.ClosureMetaClass.invokeMethod(ClosureMetaClass.java:342) ~[groovy-4.0.5.jar:4.0.5]
	at org.codehaus.groovy.runtime.callsite.PogoMetaClassSite.callCurrent(PogoMetaClassSite.java:63) ~[groovy-4.0.5.jar:4.0.5]
	at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCallCurrent(CallSiteArray.java:49) ~[groovy-4.0.5.jar:4.0.5]
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.callCurrent(AbstractCallSite.java:171) ~[groovy-4.0.5.jar:4.0.5]
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.callCurrent(AbstractCallSite.java:203) ~[groovy-4.0.5.jar:4.0.5]
	at liquibase.parser.ext.GroovyLiquibaseChangeLogParser$_getChangeLogMethodMissing_closure3.doCall(GroovyLiquibaseChangeLogParser.groovy:91) ~[liquibase-groovy-dsl-3.0.2.jar:?]
	at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:?]
	at jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[?:?]
	at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:?]
	at java.lang.reflect.Method.invoke(Method.java:566) ~[?:?]
	at org.codehaus.groovy.reflection.CachedMethod.invoke(CachedMethod.java:343) ~[groovy-4.0.5.jar:4.0.5]
	at org.codehaus.groovy.runtime.metaclass.ClosureMetaMethod.invoke(ClosureMetaMethod.java:88) ~[groovy-4.0.5.jar:4.0.5]
	at groovy.lang.MetaClassImpl.invokeMissingMethod(MetaClassImpl.java:926) ~[groovy-4.0.5.jar:4.0.5]
	at groovy.lang.MetaClassImpl.invokePropertyOrMissing(MetaClassImpl.java:1428) ~[groovy-4.0.5.jar:4.0.5]
	at groovy.lang.MetaClassImpl.doInvokeMethod(MetaClassImpl.java:1350) ~[groovy-4.0.5.jar:4.0.5]
	at groovy.lang.MetaClassImpl.invokeMethod(MetaClassImpl.java:1103) ~[groovy-4.0.5.jar:4.0.5]
	at groovy.lang.ExpandoMetaClass.invokeMethod(ExpandoMetaClass.java:1142) ~[groovy-4.0.5.jar:4.0.5]
	at groovy.lang.MetaClassImpl.invokeMethod(MetaClassImpl.java:1009) ~[groovy-4.0.5.jar:4.0.5]
	at org.codehaus.groovy.vmplugin.v8.IndyInterface.fromCache(IndyInterface.java:318) ~[groovy-4.0.5.jar:4.0.5]
	at Script1.run(Script1.groovy:1) ~[?:?]
	at Script1$run.call(Unknown Source) ~[?:?]
	at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCall(CallSiteArray.java:45) ~[groovy-4.0.5.jar:4.0.5]
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:125) ~[groovy-4.0.5.jar:4.0.5]
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:130) ~[groovy-4.0.5.jar:4.0.5]
	at liquibase.parser.ext.GroovyLiquibaseChangeLogParser.parse(GroovyLiquibaseChangeLogParser.groovy:62) ~[liquibase-groovy-dsl-3.0.2.jar:?]
	at liquibase.Liquibase.getDatabaseChangeLog(Liquibase.java:380) ~[liquibase-core-4.15.0.jar:?]
	at liquibase.Liquibase.getDatabaseChangeLog(Liquibase.java:365) ~[liquibase-core-4.15.0.jar:?]
	at liquibase.Liquibase.lambda$update$1(Liquibase.java:222) ~[liquibase-core-4.15.0.jar:?]
	at liquibase.Scope.lambda$child$0(Scope.java:180) ~[liquibase-core-4.15.0.jar:?]
	at liquibase.Scope.child(Scope.java:189) ~[liquibase-core-4.15.0.jar:?]
	at liquibase.Scope.child(Scope.java:179) ~[liquibase-core-4.15.0.jar:?]
	at liquibase.Scope.child(Scope.java:158) ~[liquibase-core-4.15.0.jar:?]
	at liquibase.Liquibase.runInScope(Liquibase.java:2414) ~[liquibase-core-4.15.0.jar:?]
	at liquibase.Liquibase.update(Liquibase.java:209) ~[liquibase-core-4.15.0.jar:?]
	at liquibase.Liquibase.update(Liquibase.java:195) ~[liquibase-core-4.15.0.jar:?]
	at liquibase.integration.spring.SpringLiquibase.performUpdate(SpringLiquibase.java:314) ~[liquibase-core-4.15.0.jar:?]
	at liquibase.integration.spring.SpringLiquibase.afterPropertiesSet(SpringLiquibase.java:269) ~[liquibase-core-4.15.0.jar:?]
	at org.springframework.boot.autoconfigure.liquibase.DataSourceClosingSpringLiquibase.afterPropertiesSet(DataSourceClosingSpringLiquibase.java:46) ~[spring-boot-autoconfigure-2.7.4.jar:2.7.4]
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.invokeInitMethods(AbstractAutowireCapableBeanFactory.java:1863) ~[spring-beans-5.3.23.jar:5.3.23]
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.initializeBean(AbstractAutowireCapableBeanFactory.java:1800) ~[spring-beans-5.3.23.jar:5.3.23]
	... 18 more
```

</p></details>

From what I can tell, there is a small difference in the way Groovy 4 finds methods from the owner and delegates.

Before, it would look at the resolve strategy, and then attempt to find the method in either the owner or the delegate. If not found in the first, it would look in the second.

Now, it "delegates" the method call to the owner or delegate directly. For instance, if the resolve strategy is set to OWNER_FIRST, it will call the method on the owner. If not found, it will then invoke the `methodMissing` method on the metaClass. For Liquibase, it causes `getChangeLogMethodMissing` to be called on the owner, which then throws the `ChangeLogParseException` because the method is _not_ on the owner - it is on the delegate. But because this exception isn't caught by Groovy (only `MissingMethodException` are caught), it just fails immediately without trying the delegate.

Don't know if that makes sense as I found it a bit hard to explain.

In our case, we can't run Liquibase in isolation as it is part of our application, which also depends on Groovy through Spock. We are trying to upgrade the whole stack to work with Groovy 4. However, I have a patch ready that fixes the problem for us.

--- a/src/main/groovy/liquibase/parser/ext/GroovyLiquibaseChangeLogParser.groovy
+++ b/src/main/groovy/liquibase/parser/ext/GroovyLiquibaseChangeLogParser.groovy
@@ -126,7 +126,7 @@ class GroovyLiquibaseChangeLogParser implements ChangeLogParser {
 
         delegate.resourceAccessor = resourceAccessor
         closure.delegate = delegate
-        closure.resolveStrategy = Closure.OWNER_FIRST
+        closure.resolveStrategy = Closure.DELEGATE_FIRST
         closure.call()
     }
 }

